Source/FakeItEasy/Core/EventHandlerArgumentProviderMap.cs

```diff
@@ -1,29 +1,24 @@
 ﻿namespace FakeItEasy.Core
 {
     using System;
+    using System.Collections.Concurrent;
     using System.Collections.Generic;
 
     internal class EventHandlerArgumentProviderMap
     {
-        private readonly Dictionary<Delegate, IEventRaiserArguments> map
-            = new Dictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
+        private readonly ConcurrentDictionary<Delegate, IEventRaiserArguments> map
+            = new ConcurrentDictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
 
         public void AddArgumentProvider(Delegate eventHandler, IEventRaiserArguments argumentProvider)
         {
-            lock (this.map)
-            {
-                this.map[eventHandler] = argumentProvider;
-            }
+            this.map[eventHandler] = argumentProvider;
         }
 
         public IEventRaiserArguments TakeArgumentProviderFor(Delegate eventHandler)
         {
-            lock (this.map)
-            {
-                var provider = this.map[eventHandler];
-                this.map.Remove(eventHandler);
-                return provider;
-            }
+            IEventRaiserArguments provider;
+            this.map.TryRemove(eventHandler, out provider);
+            return provider;

```

Comment:
Good point. It _does_ change the behaviour of the method, but not the overall behaviour. See `FakeManager.EventRule.HandleEventCall`. It will not call `RaiseEvent` (which is the only caller of this method) unless the eventCall is an event raiser, which means that the `EventHandlerArgumentProviderMap` `Contains` the `eventHandler`.

Because of this and the coupling of the argument provider registration and the raising of the event (it all happens in `robot.FellInLove += Raise.With(someEventArgs)`), there's no chance for the argument provider to be removed from the map between the `Contains` and `TakeArgumentProviderFor` being called.

I believe the code is sound as is, but if you'd like, I can make an adjustment.
Would you prefer an exception be thrown if `TryRemove` finds nothing?